common/web/gesture-recognizer/src/engine/eventSequentializationQueue.ts

```diff
@@ -0,0 +1,51 @@
+import { timedPromise } from "@keymanapp/web-utils";
+
+export class EventSequentializationQueue {
+  private queue: (() => Promise<void> | void)[];
+  private defermentPromise: Promise<void>;
+
+  constructor() {
+    this.queue = [];
+  }
+
+  private setDeferment(promise: Promise<any>) {
+    this.defermentPromise = promise;
+    promise.then(() => {
+      this.defermentPromise = null;
+      this.triggerEvent();
+    });
+  }
+
+  private async triggerEvent() {
+    while(this.queue.length > 0) {
+      const functor = this.queue.shift();
+
+      // Things break _badly_ if we don't keep the queue running if errors are triggered by the functor.
+      // It's best to ignore the error and let things play out.
+      try {
+        // Is either undefined or is a Promise.
+        const result = functor();
+        // We either wait on a manual lock (from within an InputEventEngine) or a macrotask queue wait,
+        // allowing gesture-matching microtask queue Promises to complete before proceeding.
```

Comment:
This behaviour is tailored very specifically to the users of this queue object. It feels like there are buried assumptions in the design.